File: source/com/barrybecker4/simulation/cave/model/Cave.java
package com.barrybecker4.simulation.cave.model;

import com.barrybecker4.common.math.Range;

import java.util.Random;

/**
 * Ideas for future work:
 *  - add bump map rendering and phong shading
 *   - allow continuous iterating or multiple steps at once
 *   - support mouse interaction to raise or lower the height field
 * @author Barry Becker
 */
public class Cave {

    private static final int SEED = 0;
    private static final Random RAND = new Random();

    /** a value representing the height. MAX_HEIGHT is wall, MIN_HEIGHT is floor  */
    private double[][] heightMap;
    private double floorThresh = 0.2;
    private double ceilThresh = 0.9;

    public Cave(int width, int length, double floorThresh, double ceilThresh) {
        this.floorThresh = floorThresh;
        this.ceilThresh = ceilThresh;
        heightMap = genMap(width, length);
    }

    public int getWidth() {
        return heightMap.length;
    }
    public int getLength() {
        return heightMap[0].length;
    }

    public Range getRange() {
        return new Range(floorThresh, ceilThresh);
    }
    public void setValue(int x, int y, double value) {
        heightMap[x][y] = Math.min(Math.max(value, floorThresh), ceilThresh);
    }

    public Cave createCopy() {
        Cave newCave = new Cave(getWidth(), getLength(), this.floorThresh, this.ceilThresh);
        for (int x = 0; x < getWidth(); x++) {
            for (int y = 0; y < getLength(); y++) {
                newCave.setValue(x, y, heightMap[x][y]);
            }
        }
        return newCave;
    }

    public double getValue(int x, int y) {
        return heightMap[x][y];
    }

    /**
     * @param amount the amout to change hte height by. Will never go above 1 or below 0.
     */
    public void incrementHeight(int x, int y, double amount) {
        double oldVal = heightMap[x][y];
        heightMap[x][y] = Math.max(floorThresh, Math.min(ceilThresh, oldVal + amount));
    }

    public void setFloorThresh(double floor) {
        this.floorThresh = floor;
    }

    public void setCeilThresh(double ceil) {
        this.ceilThresh = ceil;
    }

    private char getChar(int x, int y) {
        double v = heightMap[x][y];
        if (v < floorThresh) return ' ';
        else if (v < ceilThresh) return 'C';
        else return 'W';
    }

    /** generate the initial random 2D typeMap data */
    private double[][] genMap(int width, int length) {
        RAND.setSeed(SEED);
        double[][] map = new double[width][length];

        for (int x = 0; x < width; x++) {
            for (int y = 0; y < length; y++) {
                double r = RAND.nextDouble();
                map[x][y] = Math.min(Math.max(r, floorThresh), ceilThresh);
            }
        }
        return map;
    }

    public void print() {
        System.out.println(this.toString());
    }

    public String toString() {
        StringBuilder bldr = new StringBuilder();
        for (int y = 0; y < getLength(); y++) {
            for (int x = 0; x < getWidth(); x++) {
                bldr.append(getChar(x, y));
            }
            bldr.append("\n");
        }
        return bldr.toString();
    }
}

File: scala-source/com/barrybecker4/simulation/cave/model/Cave.scala
// Copyright by Barry G. Becker, 2016-2017. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.simulation.cave.model

import com.barrybecker4.math.Range
import scala.util.Random

/**
  * Data structure for representing an evolving cave system.
  * @author Barry Becker
  */
object Cave {
  private val RAND = new Random
  RAND.setSeed(0)

  /** Reseed shared RNG used by caves that omit an explicit [[scala.util.Random]] (tests, reproducible worlds). */
  def reseedRandom(seed: Long): Unit = RAND.setSeed(seed)
}

/**
  *
  * @param width width of the cave area
  * @param length length of the cave area
  * @param floorThresh values below this are in the floor
  * @param ceilThresh values above this are in the ceiling
  */
class Cave(val width: Int, val length: Int,
           var floorThresh: Double = 0.2, var ceilThresh: Double = 0.9, rnd: Random = Cave.RAND) {

  /** a value representing the height. MAX_HEIGHT is wall, MIN_HEIGHT is floor  */
  private val heightMap = Array.ofDim[Double](width, length)

  def getWidth: Int = heightMap.length
  def getLength: Int = heightMap(0).length

  def getRange: Range = Range(floorThresh, ceilThresh)

  randomInitialization()

  private def clamp(value: Double): Double =
    math.max(floorThresh, math.min(ceilThresh, value))

  def setValue(x: Int, y: Int, value: Double): Unit = {
    heightMap(x)(y) = clamp(value)
  }

  /** @return the initial random 2D typeMap data */
  def randomInitialization(): Unit = {
    for (x <- 0 until width) {
      for (y <- 0 until length) {
        val r = rnd.nextDouble()
        heightMap(x)(y) = clamp(r)
      }
    }
  }

  def createCopy: Cave = {
    val newCave = new Cave(getWidth, getLength, this.floorThresh, this.ceilThresh)
    for (x <- heightMap.indices) {
      System.arraycopy(heightMap(x), 0, newCave.heightMap(x), 0, getLength)
    }
    newCave
  }

  def getValue(x: Int, y: Int): Double = heightMap(x)(y)

  /** @param amount the amount to change the height by. Will never go above 1 or below 0. */
  def incrementHeight(x: Int, y: Int, amount: Double): Unit = {
    val oldVal = heightMap(x)(y)
    heightMap(x)(y) = clamp(oldVal + amount)
  }

  def setFloorThresh(floor: Double): Unit = { this.floorThresh = floor }
  def setCeilThresh(ceil: Double): Unit = { this.ceilThresh = ceil }

  /** @return a character representing the cave type at a specific location. Either wall, floor, or ceiling. */
  private def getChar(x: Int, y: Int): Char = {
    val v = heightMap(x)(y)
    if (v <= floorThresh) ' '
    else if (v >= ceilThresh) 'C'
    else 'W'
  }

  def print(): Unit = println(this.toString)

  override def toString: String =
    (0 until getLength)
      .map { y => (0 until getWidth).map(x => getChar(x, y)).mkString }
      .mkString("", "\n", "\n")
}
